File: services/internal/secret-broker/internal/kubernetes/provider_bootstrap_test.go
package kubernetes

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"time"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
)

// Оснастка использует настоящий producer, а не копию его labels и encoding.
// UID/RV назначает только fake Kubernetes: Python не имитирует их выдачу.
func bootstrapProviderFixture(t *testing.T) (*corev1.Secret, ProviderCredentialDescriptor) {
	t.Helper()
	producer, err := filepath.Abs("../../../../../tools/install/provider-bootstrap.py")
	if err != nil {
		t.Fatal("cannot locate bootstrap producer")
	}
	ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
	defer cancel()
	command := exec.CommandContext(ctx, "python3", "-B", "-c", `
import importlib.util, json, sys
spec = importlib.util.spec_from_file_location("provider_bootstrap", sys.argv[1])
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)
print(json.dumps(module.manifest("pacc_cleanup_Account9Z", "qa-bootstrap", sys.stdin.buffer.read())))
`, producer)
	command.Env = []string{"PATH=" + os.Getenv("PATH"), "LANG=C.UTF-8"}
	command.Stdin = strings.NewReader(providerCleanupContent)
	output, err := command.Output()
	if err != nil {
		t.Fatal("bootstrap fixture producer failed")
	}
	defer clear(output)
	var secret corev1.Secret
	if json.Unmarshal(output, &secret) != nil {
		t.Fatal("bootstrap fixture is not a Kubernetes Secret")
	}
	if secret.Namespace != "kodex-runtime" {
		t.Fatal("bootstrap namespace does not match runtime")
	}
	secret.UID = types.UID("61000000-0000-4000-8000-000000000002")
	secret.ResourceVersion = "bootstrap-7"
	digest := sha256.Sum256([]byte(providerCleanupContent))
	return &secret, ProviderCredentialDescriptor{
		SecretName: secret.Name, SecretUID: string(secret.UID),
		SecretResourceVersion: secret.ResourceVersion, ContentSHA256: hex.EncodeToString(digest[:]),
	}
}

func TestBootstrapProviderCredentialExactProducerReadback(t *testing.T) {
	t.Parallel()
	secret, descriptor := bootstrapProviderFixture(t)
	defer clearSecretData(secret)
	store, client := newTestStore(t, secret)
	store.namespace = secret.Namespace
	for range 2 {
		value, err := store.ReadProviderCredentialExact(t.Context(), providerCleanupAccountRef, descriptor)
		if err != nil || !bytes.Equal(value, []byte(providerCleanupContent)) {
			clear(value)
			t.Fatal("canonical bootstrap exact readback failed")
		}
		// Очистка полученной копии не должна менять immutable объект или повторный read.
		clear(value)
	}
	for _, action := range client.Actions() {
		if action.GetVerb() != "get" || action.GetNamespace() != secret.Namespace {
			t.Fatal("credential readback performed an unexpected action")
		}
	}
	stored, err := client.CoreV1().Secrets(secret.Namespace).Get(t.Context(), secret.Name, metav1.GetOptions{})
	if err != nil || !bytes.Equal(stored.Data[providerAuthJSONKey], []byte(providerCleanupContent)) {
		t.Fatal("readback modified the stored credential")
	}
	clearSecretData(stored)
}

func TestBootstrapProviderCredentialExactRejectsInvalidMaterialization(t *testing.T) {
	t.Parallel()
	canonical, descriptor := bootstrapProviderFixture(t)
	defer clearSecretData(canonical)
	cases := []struct {
		name   string
		change func(*corev1.Secret)
	}{
		{"digest LF", func(s *corev1.Secret) { s.Data[providerAuthSHA256Key] = append(s.Data[providerAuthSHA256Key], '\n') }},
		{"legacy publisher", func(s *corev1.Secret) { s.Labels[providerManagedByLabel] = "kodex-local-dev" }},
		{"installer publisher", func(s *corev1.Secret) { s.Labels[providerManagedByLabel] = "kodex-install" }},
		{"foreign publisher", func(s *corev1.Secret) { s.Labels[providerManagedByLabel] = "other-controller" }},
		{"missing publisher", func(s *corev1.Secret) { delete(s.Labels, providerManagedByLabel) }},
		{"wrong bootstrap version", func(s *corev1.Secret) { s.Labels[providerBootstrapLabel] = "v2" }},
		{"missing bootstrap version", func(s *corev1.Secret) { delete(s.Labels, providerBootstrapLabel) }},
		{"wrong part of", func(s *corev1.Secret) { s.Labels[providerPartOfLabel] = "other" }},
		{"wrong account", func(s *corev1.Secret) { s.Annotations[providerAccountRefAnnotation] = "pacc_otherAccount9" }},
		{"missing account", func(s *corev1.Secret) { delete(s.Annotations, providerAccountRefAnnotation) }},
		{"wrong digest annotation", func(s *corev1.Secret) { s.Annotations[providerContentSHAAnnotation] = strings.Repeat("a", 64) }},
		{"missing digest annotation", func(s *corev1.Secret) { delete(s.Annotations, providerContentSHAAnnotation) }},
		{"changed auth bytes", func(s *corev1.Secret) {
			s.Data[providerAuthJSONKey] = []byte(`{"auth_mode":"apikey","OPENAI_API_KEY":"other-synthetic"}`)
		}},
		{"wrong UID", func(s *corev1.Secret) { s.UID = types.UID("61000000-0000-4000-8000-000000000003") }},
		{"missing UID", func(s *corev1.Secret) { s.UID = "" }},
		{"wrong resource version", func(s *corev1.Secret) { s.ResourceVersion = "bootstrap-8" }},
		{"missing resource version", func(s *corev1.Secret) { s.ResourceVersion = "" }},
		{"additional data key", func(s *corev1.Secret) { s.Data["unexpected"] = []byte("synthetic") }},
		{"missing digest key", func(s *corev1.Secret) { delete(s.Data, providerAuthSHA256Key) }},
		{"mutable", func(s *corev1.Secret) { *s.Immutable = false }},
		{"missing immutable", func(s *corev1.Secret) { s.Immutable = nil }},
		{"wrong type", func(s *corev1.Secret) { s.Type = corev1.SecretTypeTLS }},
	}
	for _, scenario := range cases {
		t.Run(scenario.name, func(t *testing.T) {
			secret := canonical.DeepCopy()
			defer clearSecretData(secret)
			scenario.change(secret)
			store, client := newTestStore(t, secret)
			store.namespace = secret.Namespace
			value, err := store.ReadProviderCredentialExact(t.Context(), providerCleanupAccountRef, descriptor)
			defer clear(value)
			if !errors.Is(err, ErrProviderCredentialConflict) || len(value) != 0 {
				t.Fatal("invalid bootstrap credential did not fail closed")
			}
			for _, action := range client.Actions() {
				if action.GetVerb() != "get" {
					t.Fatal("rejection mutated the credential")
				}
			}
		})
	}
}

func TestBootstrapProviderCredentialExactRejectsWrongRequestedPins(t *testing.T) {
	t.Parallel()
	secret, descriptor := bootstrapProviderFixture(t)
	defer clearSecretData(secret)
	store, _ := newTestStore(t, secret)
	store.namespace = secret.Namespace
	for _, field := range []string{"account", "UID", "resourceVersion", "digest", "name"} {
		t.Run(field, func(t *testing.T) {
			wanted, account := descriptor, providerCleanupAccountRef
			switch field {
			case "account":
				account = "pacc_otherAccount9"
			case "UID":
				wanted.SecretUID = "61000000-0000-4000-8000-000000000003"
			case "resourceVersion":
				wanted.SecretResourceVersion = "bootstrap-8"
			case "digest":
				wanted.ContentSHA256 = strings.Repeat("b", 64)
			case "name":
				wanted.SecretName = "provider-bootstrap-absent"
			}
			value, err := store.ReadProviderCredentialExact(t.Context(), account, wanted)
			defer clear(value)
			if !errors.Is(err, ErrProviderCredentialConflict) || len(value) != 0 {
				t.Fatal("wrong requested credential pins did not fail closed")
			}
		})
	}
}

func TestBootstrapProviderCredentialCleanupExactTarget(t *testing.T) {
	t.Parallel()
	canonical, descriptor := bootstrapProviderFixture(t)
	defer clearSecretData(canonical)
	if canonical.Labels[providerCredentialLabel] != "" {
		t.Fatal("bootstrap must not claim broker credential discovery provenance")
	}
	for _, variant := range []string{"exact", "wrong UID", "wrong resourceVersion"} {
		t.Run(variant, func(t *testing.T) {
			secret := canonical.DeepCopy()
			defer clearSecretData(secret)
			store, client := newTestStore(t, secret)
			store.namespace = secret.Namespace
			wanted := descriptor
			if variant == "wrong UID" {
				wanted.SecretUID = "61000000-0000-4000-8000-000000000003"
			}
			if variant == "wrong resourceVersion" {
				wanted.SecretResourceVersion = "bootstrap-8"
			}
			receipt, err := store.CleanupProviderCredential(t.Context(), providerCleanupTaskRef,
				providerCleanupAccountRef, providerCleanupGeneration, wanted)
			if variant == "exact" {
				if err != nil || receipt.TerminalReceipt == "" || receipt.ProducedCredential != nil {
					t.Fatal("exact bootstrap cleanup did not produce a terminal receipt")
				}
				assertProviderSecretDeletePreconditions(t, client, descriptor.SecretName, descriptor.SecretUID, descriptor.SecretResourceVersion)
				fence, getErr := client.CoreV1().Secrets(secret.Namespace).Get(t.Context(), secret.Name, metav1.GetOptions{})
				if getErr != nil || len(fence.Data[providerAuthJSONKey]) != 0 {
					t.Fatal("cleanup retained credential material")
				}
				if _, parseErr := store.parseProviderCleanupFence(fence, "CREDENTIAL", providerCleanupAccountRef); parseErr != nil {
					t.Fatal("cleanup did not fence the bootstrap name")
				}
				return
			}
			if providerSecretActionCount(client, "delete", secret.Name) != 0 {
				t.Fatal("stale cleanup deleted bootstrap credential")
			}
			if variant == "wrong resourceVersion" && !errors.Is(err, ErrProviderCredentialCleanupConflict) {
				t.Fatal("wrong resourceVersion did not conflict")
			}
			// Смена UID сообщает владельцу отдельную replacement revision, не удаляет её.
			if variant == "wrong UID" && (err != nil || receipt.ProducedCredential == nil || *receipt.ProducedCredential != descriptor) {
				t.Fatal("replacement descriptor was not preserved")
			}
			value, readErr := store.ReadProviderCredentialExact(t.Context(), providerCleanupAccountRef, descriptor)
			defer clear(value)
			if readErr != nil || !bytes.Equal(value, []byte(providerCleanupContent)) {
				t.Fatal("stale cleanup changed the current credential")
			}
		})
	}
}

func TestBootstrapProviderCredentialExactPreservesCanonicalPublishers(t *testing.T) {
	t.Parallel()
	for _, manager := range []string{providerSecretBrokerManager, providerRuntimeManager} {
		t.Run(manager, func(t *testing.T) {
			secret, descriptor := providerCleanupFixture(manager)
			defer clearSecretData(secret)
			store, _ := newTestStore(t, secret)
			value, err := store.ReadProviderCredentialExact(t.Context(), providerCleanupAccountRef, descriptor)
			defer clear(value)
			if err != nil || !bytes.Equal(value, []byte(providerCleanupContent)) {
				t.Fatal("existing canonical publisher was rejected")
			}
			wrong := descriptor
			wrong.SecretResourceVersion = "different-version"
			denied, err := store.ReadProviderCredentialExact(t.Context(), providerCleanupAccountRef, wrong)
			defer clear(denied)
			if !errors.Is(err, ErrProviderCredentialConflict) || len(denied) != 0 {
				t.Fatal("existing publisher bypassed exact descriptor checks")
			}
		})
	}
}

File: scripts/tests/provider-bootstrap-test.py
#!/usr/bin/env python3
"""Изолированная оснастка publisher: Kubernetes/SQL здесь синтетические."""
import base64
import copy
import importlib.util
import json
from pathlib import Path
import tempfile
import unittest

ROOT = Path(__file__).resolve().parents[2]
spec = importlib.util.spec_from_file_location("publisher", ROOT / "tools/install/provider-bootstrap.py")
p = importlib.util.module_from_spec(spec)
spec.loader.exec_module(p)
RAW = b'{"auth_mode":"apikey","OPENAI_API_KEY":"synthetic-provider-material"}'
ACCOUNT = "pacc_bootstrap_fixture"
KEY = "default-openai-codex"


def stored(raw=RAW):
    value = p.manifest(ACCOUNT, KEY, raw)
    value["metadata"].update(uid="10000000-0000-4000-8000-000000000001", resourceVersion="7")
    return value


class Memory(p.Operator):
    def __init__(self, secret):
        super().__init__("synthetic")
        self.secrets = {secret["metadata"]["name"]: copy.deepcopy(secret)}
        self.state = dict(accountRef=ACCOUNT, version=1, state="AUTHORIZED", enabled=True,
                          credentialRef="pcr_initialfixture", credential=p.descriptor(secret))
        self.effects = []
        self.conflict = False
        self.unknown = False

    def snapshot(self, key):
        return copy.deepcopy(self.state)

    def get(self, kind, name, namespace="kodex-runtime", absent=False):
        return copy.deepcopy(self.secrets.get(name))

    def call(self, argv, payload=None):
        if argv[0] != "create":
            raise AssertionError("unexpected effect")
        value = json.loads(payload)
        name = value["metadata"]["name"]
        assert name not in self.secrets
        value["metadata"].update(uid="20000000-0000-4000-8000-000000000001", resourceVersion="8")
        self.secrets[name] = value
        self.effects.append("create")
        if self.unknown:
            raise p.Rejected("unknown")
        return b""

    def sql(self, name, values):
        assert name == "reconcile-provider-account.sql"
        self.effects.append("cas")
        if self.conflict:
            raise p.Rejected("stale")
        assert values["expected_version"] == self.state["version"]
        self.state.update(version=self.state["version"] + 1, credentialRef=values["credential_ref"], credential={
            "secretName": values["secret_name"], "secretUID": values["secret_uid"],
            "secretResourceVersion": values["secret_resource_version"], "contentSHA256": values["content_sha256"]})
        return True

    def metadata(self, snapshot):
        assert snapshot["credential"] == self.state["credential"]
        self.effects.append("metadata")


class Tests(unittest.TestCase):
    def legacy(self):
        value = stored()
        value["metadata"]["name"] = "runtime-provider-openai-default-r1"
        value["metadata"]["labels"] = {"app.kubernetes.io/managed-by": "kodex-install"}
        value["data"]["auth.sha256"] = base64.b64encode((p.digest(RAW) + "\n").encode()).decode()
        return value

    def test_exact_canonical_and_rejections(self):
        value = stored()
        pins = p.descriptor(value)
        self.assertEqual(p.validate(value, pins, ACCOUNT, KEY), (RAW, True))
        for field, replacement in (("secretUID", "20000000-0000-4000-8000-000000000002"),
                                   ("secretResourceVersion", "9"), ("contentSHA256", "b" * 64)):
            with self.subTest(field=field), self.assertRaises(p.Rejected):
                p.validate(value, {**pins, field: replacement}, ACCOUNT, KEY)
        for edit in (lambda x: x["metadata"]["labels"].update({"app.kubernetes.io/managed-by": "unknown"}),
                     lambda x: x["metadata"]["annotations"].update({p.PREFIX + "account-ref": "pacc_foreignfixture"}),
                     lambda x: x["data"].update({"auth.sha256": base64.b64encode((p.digest(RAW) + "\n").encode()).decode()})):
            changed = copy.deepcopy(value)
            edit(changed)
            with self.assertRaises(p.Rejected):
                p.validate(changed, pins, ACCOUNT, KEY)

    def test_legacy_is_only_migration_input_and_preserved(self):
        legacy = self.legacy()
        with self.assertRaises(p.Rejected):
            p.validate(legacy, p.descriptor(legacy), ACCOUNT, KEY)
        operator = Memory(legacy)
        operator.reconcile(KEY)
        self.assertEqual(operator.secrets[legacy["metadata"]["name"]], legacy)
        self.assertEqual(operator.effects, ["create", "cas", "metadata"])
        operator.effects.clear()
        operator.reconcile(KEY)
        self.assertEqual(operator.effects, ["metadata"])
        self.assertEqual(operator.state["version"], 2)

    def test_current_selection_not_downgraded_by_old_file(self):
        operator = Memory(stored())
        operator.reconcile(KEY, "/nonexistent-stale-input", preserve_current=True)
        self.assertEqual(operator.effects, ["metadata"])

    def test_stale_cas_preserves_old_revision_and_created_object(self):
        operator = Memory(self.legacy())
        before = copy.deepcopy(operator.state)
        operator.conflict = True
        with self.assertRaises(p.Rejected):
            operator.reconcile(KEY)
        self.assertEqual(operator.state, before)
        self.assertEqual(len(operator.secrets), 2)
        self.assertNotIn("metadata", operator.effects)

    def test_lost_create_ack_next_attempt_reads_existing_without_second_create(self):
        operator = Memory(self.legacy())
        operator.unknown = True
        with self.assertRaises(p.Rejected):
            operator.reconcile(KEY)
        operator.unknown = False
        operator.reconcile(KEY)
        self.assertEqual(operator.effects.count("create"), 1)

    def test_explicit_import_uses_private_material(self):
        operator = Memory(stored())
        with tempfile.TemporaryDirectory() as directory:
            file = Path(directory) / "auth.json"
            file.write_bytes(RAW.replace(b"material", b"new-material"))
            file.chmod(0o644)
            with self.assertRaises(p.Rejected):
                operator.reconcile(KEY, str(file))
            file.chmod(0o600)
            operator.reconcile(KEY, str(file))
            self.assertEqual(operator.state["version"], 2)


if __name__ == "__main__":
    unittest.main()
